File: server/src/main/java/com/cyberbot/bomberman/server/ServerService.java
package com.cyberbot.bomberman.server;

import com.cyberbot.bomberman.core.models.net.InvalidPacketFormatException;
import com.cyberbot.bomberman.core.models.net.SerializationUtils;
import com.cyberbot.bomberman.core.models.net.packets.LobbyCreateRequest;
import com.cyberbot.bomberman.core.models.net.packets.LobbyCreateResponse;
import com.cyberbot.bomberman.core.models.net.packets.PlayerSnapshotPacket;
import com.cyberbot.bomberman.core.utils.Utils;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.util.HashMap;
import java.util.Map;

import static com.cyberbot.bomberman.core.utils.Constants.LOBBY_ID_LENGTH;

public class ServerService implements Runnable {
    private final int port;
    private final Map<String, Session> sessions = new HashMap<>();
    private DatagramSocket socket;

    public ServerService(int port) {
        this.port = port;
    }

    @SuppressWarnings("InfiniteLoopStatement")
    @Override
    public void run() {
        try {
            socket = new DatagramSocket(port);
            for (; ; ) {
                byte[] buffer = new byte[1024];
                DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
                socket.receive(packet);

                ClientConnection client = ClientConnection.fromDatagramPacket(packet);

                Object o = SerializationUtils.INSTANCE.deserialize(buffer, packet.getLength(), 0);

                if (o instanceof PlayerSnapshotPacket) {
                    sessions.values().stream()
                        .filter(it -> it.hasClient(client))
                        .findFirst()
                        .ifPresent(session -> session.onSnapshot(client, (PlayerSnapshotPacket) o));
                } else if (o instanceof LobbyCreateRequest) {
                    String name = ((LobbyCreateRequest) o).getName();
                    Session session = new Session(name, this);
                    session.addClient(client);

                    String sessionId;
                    do {
                        sessionId = Utils.generateLobbyId(LOBBY_ID_LENGTH);
                    } while (sessions.containsKey(sessionId));

                    sessions.put(sessionId, session);

                    LobbyCreateResponse response = new LobbyCreateResponse(name, sessionId);

                    byte[] bytes = SerializationUtils.INSTANCE.serialize(response);
                    send(client, bytes, bytes.length);
                }
            }
        } catch (IOException | InvalidPacketFormatException e) {
            e.printStackTrace();
        }
    }

    public void send(DatagramPacket packet) throws IOException {
        socket.send(packet);
    }

    private void send(ClientConnection client, byte[] data, int length) throws IOException {
        DatagramPacket packet = new DatagramPacket(data, 0, length, client.getAddress(), client.getPort());
        socket.send(packet);
    }
}

File: server/src/main/java/com/cyberbot/bomberman/server/ServerService.kt
package com.cyberbot.bomberman.server

import com.cyberbot.bomberman.core.models.items.Inventory
import com.cyberbot.bomberman.core.models.net.data.PlayerData
import com.cyberbot.bomberman.core.models.net.packets.*
import com.cyberbot.bomberman.core.utils.Utils
import java.io.IOException
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.util.*
import java.util.concurrent.ThreadLocalRandom
import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import kotlin.concurrent.schedule


class ServerService(
    private val port: Int,
    private val maxLobbyCount: Int = 5,
    private val lobbyIdLength: Int = 5,
    private val maxPlayersPerLobby: Int = 4
) : ClientController, Runnable {
    private val sessions = HashMap<String, SessionService>()
    private val clients = HashMap<Client, ClientControlService>()
    private val lobbies = HashMap<String, Lobby>()

    @Volatile
    private var running = false
    private var serverSocket = ServerSocket()

    override fun run() {
        try {
            running = true
            serverSocket.bind(InetSocketAddress(port))
            while (running) {
                val client = serverSocket.accept()
                Thread(ClientControlService(client, this)).start()
            }
        } catch (e: IOException) {
            e.printStackTrace()
            running = false
        }
    }

    override fun onClientRegister(
        request: ClientRegisterRequest,
        service: ClientControlService
    ): ClientRegisterResponse {
        request.nick ?: return ClientRegisterResponse(false)
        val client = createClient(request.nick!!)

        client?.also {
            clients[it] = service
        }

        return ClientRegisterResponse(client != null, client)
    }

    override fun onLobbyCreate(request: LobbyCreateRequest, client: Client): LobbyCreateResponse {
        return if (lobbies.size < maxLobbyCount) {
            val lobby = createLobby(client)
            val id = lobby.id ?: throw RuntimeException("Created lobby missing id")
            lobbies[id] = lobby

            LobbyCreateResponse(true, id)
        } else {
            LobbyCreateResponse(false)
        }
    }

    override fun onLobbyJoin(request: LobbyJoinRequest, client: Client): LobbyJoinResponse? {
        val lobby = lobbies[request.id] ?: return LobbyJoinResponse(false)

        return if (lobby.clients.size < maxPlayersPerLobby) {
            lobby.clients.add(client)

            // Quick and dirty way to first send the response to a client and only later the update packet
            Timer().schedule(100) {
                notifyLobbyChange(lobby)
            }

            LobbyJoinResponse(true)
        } else {
            LobbyJoinResponse(false)
        }
    }

    override fun onGameStart(request: GameStartRequest, client: Client) {
        val lobby = lobbies.values.firstOrNull { it.ownerId == client.id } ?: return
        val lobbyId = lobby.id ?: throw RuntimeException("Lobby in lobbies without id")

        val session = SessionService()
        sessions[lobbyId] = session

        lobby.clients.forEachIndexed { i, c ->
            val id = c.id ?: throw RuntimeException("Client without id")
            val data = PlayerData(id, Session.getPlayerSpawnPosition(i), Inventory(), i)

            session.addClient(c.id!!, data)
            // Clients has to contain a client that's present in a lobby
            clients[c]!!.sendPacket(GameStart(session.port, data))
        }

        Thread(session).start()
    }

    private fun notifyLobbyChange(lobby: Lobby) {
        val lobbyUpdate = LobbyUpdate(System.currentTimeMillis(), Lobby.stripIds(lobby))
        lobby.clients.map { clients[it] }.forEach { it?.sendPacket(lobbyUpdate) }
    }

    private fun createLobby(owner: Client): Lobby {
        // Remove any lobbies previously created by this client
        lobbies.entries.removeIf { it.value.ownerId == owner.id }

        var id: String

        do {
            id = Utils.generateLobbyId(lobbyIdLength)
        } while (lobbies.keys.contains(id))

        return Lobby(id, owner.id, ArrayList())
    }

    private fun createClient(nick: String): Client? {
        val clients = clients.map { it.key }
        if (clients.map { it.nick }.contains(nick)) {
            return null
        }

        val ids = clients.map { it.id }

        var id: Long
        do {
            id = ThreadLocalRandom.current().nextLong()
        } while (ids.contains(id))

        return Client(id, nick)
    }
}